protocol/src/types/cstring.rs

```diff
@@ -0,0 +1,61 @@
+use std::ffi::CString;
+use std::io::prelude::{Read, Write};
+use {hint, Error, Parcel, Settings};
+
+impl Parcel for CString {
+    const TYPE_NAME: &'static str = "CString";
+
+    fn read_field(
+        read: &mut dyn Read,
+        settings: &Settings,
+        _hints: &mut hint::Hints,
+    ) -> Result<Self, Error> {
+        let mut result = Vec::new();
+        loop {
+            let c: u8 = Parcel::read(read, settings)?;
+            if c == 0x00 {
+                break;
```

Comment:
I suspect this `break` needs to be underneath `result.push(c)` - currently, this logic will collect every byte except the null terminator.

The [ffi::CString::new](file:///home/dylan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/doc/rust/html/std/ffi/struct.CString.html#method.new) method requires a null terminator on the input buffer.

Reply:
Taken from the docs (bold is mine):

This function will consume the provided data and use the underlying bytes to construct a new string, ensuring that there is a trailing 0 byte. This trailing 0 byte will be appended by this function; **the provided data should not contain any 0 bytes in it.**

Reply:
Good catch, I also looked at the docs but somehow took the complete opposite meaning from what the docs intended. I re-read it as you've restated and indeed, the null byte should not be passed.